plugins/synaptics-rmi/fu-synaptics-rmi-device.c

```diff
@@ -0,0 +1,1070 @@
+				     80, FU_DUMP_FLAGS_NONE);
+	}
+
+	return g_steal_pointer (&buf);
+}
+
+gboolean
+fu_synaptics_rmi_device_write (FuSynapticsRmiDevice *self, guint16 addr, GByteArray *req, GError **error)
+{
+	FuSynapticsRmiDevicePrivate *priv = GET_PRIVATE (self);
+	guint8 len = 0x0;
+	g_autoptr(GByteArray) buf = g_byte_array_new ();
+
+	/* check size */
+	if (req != NULL) {
+		if (req->len > 0xff) {
+			g_set_error_literal (error,
+					     FWUPD_ERROR,
+					     FWUPD_ERROR_INTERNAL,
+					     "data to write was too long");
+			return FALSE;
+		}
+		len = req->len;
+	}
+
+	/* report */
+	fu_byte_array_append_uint8 (buf, RMI_WRITE_REPORT_ID);
+
+	/* length */
+	fu_byte_array_append_uint8 (buf, len);
+
+	/* address */
+	fu_byte_array_append_uint16 (buf, addr, G_LITTLE_ENDIAN);
+
+	/* optional data */
+	if (req != NULL)
+		g_byte_array_append (buf, req->data, req->len);
+
+	/* pad out to 21 bytes for some reason */
+	for (guint i = buf->len; i < 21; i++)
+		fu_byte_array_append_uint8 (buf, 0x0);
+	if (g_getenv ("FWUPD_SYNAPTICS_RMI_VERBOSE") != NULL) {
+		fu_common_dump_full (G_LOG_DOMAIN, "DeviceWrite", buf->data, buf->len,
+				     80, FU_DUMP_FLAGS_NONE);
+	}
+
+	return fu_io_channel_write_byte_array (priv->io_channel, buf, RMI_DEVICE_DEFAULT_TIMEOUT,
+					       FU_IO_CHANNEL_FLAG_SINGLE_SHOT |
+					       FU_IO_CHANNEL_FLAG_USE_BLOCKING_IO,
+					       error);
+}
+
+static gboolean
+fu_synaptics_rmi_device_set_rma_page (FuSynapticsRmiDevice *self, guint8 page, GError **error)
+{
+	FuSynapticsRmiDevicePrivate *priv = GET_PRIVATE (self);
+	g_autoptr(GByteArray) req = g_byte_array_new ();
+
+	/* same */
+	if (priv->page == page)
+		return TRUE;
+
+	/* write */
+	fu_byte_array_append_uint8 (req, page);
+	if (!fu_synaptics_rmi_device_write (self, RMI_DEVICE_PAGE_SELECT_REGISTER, req, error)) {
+		priv->page = -1;
+		return FALSE;
+	}
+	priv->page = page;
+	return TRUE;
+}
+
+gboolean
+fu_synaptics_rmi_device_reset (FuSynapticsRmiDevice *self, GError **error)
+{
+	FuSynapticsRmiDevicePrivate *priv = GET_PRIVATE (self);
+	g_autoptr(GByteArray) req = g_byte_array_new ();
+
+	g_debug ("resetting...");
+	fu_byte_array_append_uint8 (req, RMI_F01_CMD_DEVICE_RESET);
+	if (!fu_synaptics_rmi_device_write (self, priv->f01->command_base, req, error))
+		return FALSE;
+	g_usleep (1000 * RMI_F01_DEFAULT_RESET_DELAY_MS);
+	g_debug ("reset completed");
+	return TRUE;
+}
+
+static gboolean
+fu_synaptics_rmi_device_scan_pdt (FuSynapticsRmiDevice *self, GError **error)
+{
+	FuSynapticsRmiDevicePrivate *priv = GET_PRIVATE (self);
+	guint interrupt_count = 0;
+
+	/* clear old list */
+	g_ptr_array_set_size (priv->functions, 0);
+
+	/* scan pages */
+	for (guint page = 0; page < RMI_DEVICE_MAX_PAGE; page++) {
+		guint page_start = RMI_DEVICE_PAGE_SIZE * page;
+		guint pdt_start = page_start + RMI_DEVICE_PAGE_SCAN_START;
+		guint pdt_end = page_start + RMI_DEVICE_PAGE_SCAN_END;
+
+		/* set page */
+		if (!fu_synaptics_rmi_device_set_rma_page (self, page, error))
```

Comment:
I would think you want to prefix the page you tried to set and failed for debugging

Reply:
Agree, thanks.